processing/src/main/java/io/druid/segment/StringDimensionQueryHelper.java

```diff
@@ -0,0 +1,443 @@
+          } else {
+            for (int i = 0; i < size; ++i) {
+              if (predicate.apply(selector.lookupName(row.get(i)))) {
+                return true;
+              }
+            }
+            return false;
+          }
+        }
+      };
+    }
+  }
+
+  public void hashRow(Object dimSelector, Hasher hasher)
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    final IndexedInts row = selector.getRow();
+    final int size = row.size();
+    // nothing to add to hasher if size == 0, only handle size == 1 and size != 0 cases.
+    if (size == 1) {
+      final String value = selector.lookupName(row.get(0));
+      hasher.putUnencodedChars(value != null ? value : CardinalityAggregator.NULL_STRING);
+    } else if (size != 0) {
+      final String[] values = new String[size];
+      for (int i = 0; i < size; ++i) {
+        final String value = selector.lookupName(row.get(i));
+        values[i] = value != null ? value : CardinalityAggregator.NULL_STRING;
+      }
+      // Values need to be sorted to ensure consistent multi-value ordering across different segments
+      Arrays.sort(values);
+      for (int i = 0; i < size; ++i) {
+        if (i != 0) {
+          hasher.putChar(CardinalityAggregator.SEPARATOR);
+        }
+        hasher.putUnencodedChars(values[i]);
+      }
+    }
+  }
+
+  public void hashValues(Object dimSelector, HyperLogLogCollector collector)
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    for (final Integer index : selector.getRow()) {
+      final String value = selector.lookupName(index);
+      collector.add(CardinalityAggregator.hashFn.hashUnencodedChars(value == null ? CardinalityAggregator.NULL_STRING : value).asBytes());
+    }
+  }
+
+  @Override
+  public int getGroupingKeySize()
+  {
+    return Ints.BYTES;
+  }
+
+  @Override
+  public Comparator<byte[]> getGroupingKeyByteComparator()
+  {
+    return GROUPING_KEY_COMPARATOR;
+  }
+
+  @Override
+  public void readDimValueFromGroupingKey(
+      Map<String, Object> theEvent, String outputName, Object dimSelector, ByteBuffer keyBuffer
+  )
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    final int dimVal = keyBuffer.getInt();
+    if (dimVal != GROUP_BY_MISSING_VALUE) {
+      theEvent.put(outputName, selector.lookupName(dimVal));
+    }
+  }
+
+  @Override
+  public List<ByteBuffer> addDimValuesToGroupingKey(
+      Object selector, ByteBuffer key, Function<ByteBuffer, List<ByteBuffer>> updateValuesFn
+  )
+  {
+    List<ByteBuffer> unaggregatedBuffers = null;
+    final DimensionSelector dimSelector = (DimensionSelector) selector;
+    final IndexedInts row = dimSelector.getRow();
+    if (row == null || row.size() == 0) {
+      ByteBuffer newKey = key.duplicate();
+      newKey.putInt(GROUP_BY_MISSING_VALUE);
+      unaggregatedBuffers = updateValuesFn.apply(newKey);
+    } else {
+      for (Integer dimValue : row) {
```

Comment:
Could you please avoid boxing and use `IntIterator` (`IndexedInts` is `IntIterable`) or indexed iteration?